Expanded `PublicKeyCredentialCreationOptions`

Joride opened the pull request:
Hey folks,

This might be a change that brings changes to clients using the library, so please review carefully. Maybe this one requires a version-change? We could opt to maintin existing behaviour, although I would be careful with that option as well, as effectively clients are currently using an incomplete implementation of the spec? Open to hearing what direction to take and adjust as necessary.

I added some code so the options returned from beginRegistration now contain more fields from the spec.
Without these changes, I noticed that on Android, passkey-authentication could be used, however, after signup, the passkey did not seem to be stored anywhere. And so if a user was logged out, they could not sign in again with the created passkey.

With these changes, that issue is now resolved.

Joride committed:
- expanded `PublicKeyCredentialCreationOptions` with more options from the spec
- updated implementation of `beginRegistration(..)` supply milliseconds to a new initializer
N.b. `beginRegistration(..)` now returns different fields (and so different JSON).

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -38,11 +62,9 @@ public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
     /// preferred.
     public let publicKeyCredentialParameters: [PublicKeyCredentialParameters]
 
-    /// A time, in seconds, that the caller is willing to wait for the call to complete. This is treated as a
+    /// A time, in milliseconds, that the caller is willing to wait for the call to complete. This is treated as a
     /// hint, and may be overridden by the client.
-    ///
-    /// - Note: When encoded, this value is represented in milleseconds as a ``UInt32``.
-    public let timeout: Duration?
+    public let timeoutInMilliseconds: Int64?
```
Note that this was intentional so we expose a more Swift-native API for timing, and the struct encodes with milliseconds for easy integration with JS.

Joride replied:
Got it, makes sense. I'll adjust that one.

Any thoughts on how to introduce this change, given that it might change behaviour for those who are relying on the current implementation? My thoughts:

* A version change, I think this is quite a critical part of the lib, so might be the proper course. This would require immediate action from clients when they update the package.
* I could come up with some layer of abstraction (maybe a protocol or a class-cluster/factory like approach) that provides current behaviour by default, and the extended behaviour when some option is specified or a specific initialiser is called.  This would require zero effort from clients, but complicates the code base a little bit and keeps essentially an incompletely implemented spec alive.
* A combination of both above, where the new default is the extended version, but current version can still be requested as well. This would require little, but non-zero effort of clients using the current version when updating the library. This could be mitigated by changing some function signature (`beginRegistration ` comes to mind) to take an option so that when the library gets updated, clients will _have_ to specify which behaviour the want

dimitribouniol replied:
The package is in prerelease, so breaking changes can still be made (and probably should if they improve the API). That said, not sure what about these changes (unless there are more) would incur breaking changes — it mostly seems like a collection of added properties more than anything, all of which could have sensible defaults?

Joride replied:
Yes, that is what I am thinking as well. 
However: maybe, some implementations were relaying for example on the effect on android that I found undesirable: users being able to authenticate once, but unable to re-use that credential on subsequent logins.

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -21,6 +21,30 @@ import Foundation
 /// - SeeAlso: https://www.w3.org/TR/webauthn-2/#dictionary-makecredentialoptions
 public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
+    
+    init(challenge: [UInt8], 
```
Not sure if this was meant to be public, but the internal initializer should be synthesized automatically based on default values on members. If you means to make this public, it would be better in an extension.

dimitribouniol replied:
Also, note that `PublicKeyCredentialCreationOptions` is not intended to be instanciated directly — you are expected to request one from the manager, so many of these options likely belong there instead.

Joride replied:
Fair point, I'll adjust this one to be in line with your suggestions.

Joride replied:
Updated in `3a2815e1dd3bdd07bbbb57c311eac272da05bb75`

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -55,17 +77,87 @@ public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
+    }
+    
+    let hints: [Hint]
+    enum Hint: String, Encodable
```
Rather than an enum, which can never take new values after the library gets a public release, please use UnreferencedStringEnumeration: https://github.com/swift-server/webauthn-swift/blob/75c32a7730188646dfae7c44adcba30242bdf220/Sources/WebAuthn/Ceremonies/Shared/AuthenticatorAttachment.swift#L21

Joride replied:
Where can I find out more about enums being problematic after public releases? I would like to learn more about this.
Meanwhile, I'll adjust according to your suggestions.

dimitribouniol replied:
The problem is described in the motivation section here: https://forums.swift.org/t/pitch-non-frozen-enumerations/68373

Joride replied:
Ok, thank you for sharing, I'll read up on that, never thought of enums being risky 😃

Joride replied:
Done as part of `5f8974df64c68188297a6841e77df797bb11ed70`

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -55,17 +77,87 @@ public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
+    let authenticatorSelection: AuthenticatorSelection
+    struct AuthenticatorSelection: Encodable
+    {
+        enum ResidentKey: String, Encodable
+        {
+            case required = "required"
+            case preferred = "preferred"
+            case discouraged = "discouraged"
+        }
+        
+        enum UserVerification: String, Encodable
+        {
+            case required = "required"
+            case preferred = "preferred"
+            case discouraged = "discouraged"
+        }
```
Same for these, please use `UnreferencedStringEnumeration`

Joride replied:
Done as part of `5f8974df64c68188297a6841e77df797bb11ed70`

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -55,17 +77,87 @@ public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
+    
+    struct Extensions: Encodable
+    {
+        let credProps: Bool
```
Could we use a CodingKeys declaration to give this type a nicer name in swift? Without looking into the spec, I have no clue what this refers to.

Joride replied:
Yeah that is a good idea to use codingkeys to have a better name in the library for this. Currently I absolutely have not enough understanding of the spec to be able to convert the below statement into something more descriptive. Any suggestions?

> WebAuthn Extension Identifiers defines the initial set of extensions to be registered in the [IANA Registry](https://ldapwiki.com/wiki/Wiki.jsp?page=IANA%20Registry) "WebAuthn Extension Identifier" registry established by [WebAuthn-Registries](https://ldapwiki.com/wiki/Wiki.jsp?page=WebAuthn-Registries).

These [MAY](https://ldapwiki.com/wiki/Wiki.jsp?page=MAY) be implemented by [User-agents](https://ldapwiki.com/wiki/Wiki.jsp?page=User-agent) targeting broad interoperability.

dimitribouniol replied:
https://www.w3.org/TR/webauthn-2/#sctn-authenticator-credential-properties-extension

I would suggest `credentialPropertiesExtension`, set to a `CredentialPropertiesExtension.RequestOptions` type with `.requested` and `.ignore` options that render the appropriate value to the encoded output (either `true`, or the key is skipped)

Joride replied:
Not sure I follow. Conceptually, do you mean something like the following (except instead of enum I would use `UnreferencedStringEnumeration`):

```
let credentialPropertiesExtension: CredentialPropertiesExtension
struct CredentialPropertiesExtension: Encodable
{
    let options: RequestOptions
    
    enum RequestOptions: Encodable
    {
        case requested
        case ignored
    }
}
```

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -55,17 +77,87 @@ public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
+    
+    let extensions: Extensions
+    
+    struct Extensions: Encodable
+    {
```
Nit, based on formatting in the rest of the repo"
```suggestion
    struct Extensions: Encodable {
```

dimitribouniol replied:
Also, it might make a lot more sense to make dedicated public types for all of these in their own files, so we can properly model them

Joride replied:
Ah yeah, I'm a bit of an outlier with how I format my braces, used to having a linter 'fix' that on a pre-commit hook 😃 .

About the public types: I was thinking that could be considered clutter, as those types are not really needed (yet?) anywhere outside of this struct?
Happy to make the change of course, but: are you sure?

dimitribouniol replied:
Just a suggestion from my part, as I'm not an actual maintainer, but having seen the spec in its entirety, I think its warranted, plug it keeps this file highly focused.

Joride replied:
> Also, it might make a lot more sense to make dedicated public types for all of these in their own files, so we can properly model them in the rest of the code

Done as part of `5f8974df64c68188297a6841e77df797bb11ed70`. It seems the rest of the repo keeps the types mostly in the same file, so in keeping with that, the types are no dedicated public, but still live in the same file.

Joride committed:
changed parameter from representing ms to representing seconds, in line with
a suggestion from dimitribouniol: https://github.com/swift-server/webauthn-swift/pull/76/files/3d084c20a0a8753b8a6184490bff03810fc8a72d#r1721784750

dimitribouniol commented on Sources/WebAuthn/WebAuthnManager.swift:
```diff
@@ -69,7 +69,7 @@ public struct WebAuthnManager: Sendable {
             relyingParty: .init(id: configuration.relyingPartyID, name: configuration.relyingPartyName),
             publicKeyCredentialParameters: publicKeyCredentialParameters,
-            timeout: timeout,
+            timeout: (timeout?.milliseconds ?? 0) / 1000,
```
This should still be unnecessary?

Joride replied:
Yes, updated in commit `ee000536baa6801d50b9c4a1242f43a9be1d71f1`

Joride committed:
reverting to using seconds on `PublicKeyCredentialCreationOptions`' for timeout, of type `Duration`
Follow up on this discussion: https://github.com/swift-server/webauthn-swift/pull/76#discussion_r1721940603

Joride committed:
removed custom initializer from `PublicKeyCredentialCreationOptions`, moved defaults to `beginRegistration(..)` on `WebAuthnManager`
see per suggestion: https://github.com/swift-server/webauthn-swift/pull/76#discussion_r1721803173

Joride committed:
- dedicated public types for new types
- removed some types that turned out to be redundant with existing ones
- switched from using enums to using UnreferencedStringEnumeration

addresses these discussion points
https://github.com/swift-server/webauthn-swift/pull/76#discussion_r1721799400
https://github.com/swift-server/webauthn-swift/pull/76#discussion_r1721792482
https://github.com/swift-server/webauthn-swift/pull/76#discussion_r1721794102

Joride committed:
Merge branch 'main' into PublicKeyCredentialCreationOptions-expansion

Joride commented:
Just checking in on this one, as I'm getting closer to going live using this package. 
Any (rough) e.t.a. on when this will be merged in? Any pending adjustments maybe?

thoven87 commented:
@dimitribouniol @Joride are they any particular reasons why this PR can't be merged? I wanted to build on this PR to expose aaguid, aaid so that relying party can filter passkey authenticators based on the fido [metadata blob](https://fidoalliance.org/metadata/)

Joride commented:
From my perspective this is Ok from a functionality perspective. I've been using this change now for a few months, haven't run into any issues.

I think I addressed the various comments and suggestions, so I would be happy to see it merged.

dimitribouniol commented:
I'll take a closer look reviewing it tomorrow morning when I get some time.

dimitribouniol commented on Sources/WebAuthn/Ceremonies/Registration/PublicKeyCredentialCreationOptions.swift:
```diff
@@ -40,13 +41,27 @@ public struct PublicKeyCredentialCreationOptions: Encodable, Sendable {
     public let attestation: AttestationConveyancePreference
+    
+    /// Use this enumeration to communicate hints to the user-agent about how a request may be best completed. These hints are not requirements, and do not bind the user-agent, but may guide it in providing the best experience by using contextual information that the Relying Party has about the request.
+    /// https://www.w3.org/TR/webauthn-3/#enum-hints
```
Minor nit, but would be good to follow this convention for references:
https://github.com/swift-server/swift-webauthn/blob/0f31a513fe05fed6e7db98350e1c99bb93bfd295/Sources/WebAuthn/Ceremonies/Shared/AuthenticatorAttestationGloballyUniqueID.swift#L21

dimitribouniol reviewed (changes requested):
Changes notwithstanding, I'll need to first make a PR that fixes testing before this can go in.

dimitribouniol committed:
Merge branch 'main' into PublicKeyCredentialCreationOptions-expansion
